File: go/restaurants/suzies/suzies.go
package suziesRestaurant

import (
	"io"
	"regexp"
	"strings"

	"cloudtalk/model"

	"github.com/PuerkitoBio/goquery"
)

type suzies struct {
	url   string
	title string
}

func (a suzies) GetUrl() string {
	return a.url
}

func (a suzies) Parse(body io.ReadCloser) (*model.Restaurant, error) {
	doc, err := goquery.NewDocumentFromReader(body)

	if err != nil {
		return nil, err
	}

	var dayOffer []model.Day

	doc.Find("#weekly-menu .day").Each(func(i int, s *goquery.Selection) {

		title := s.Find("h4").Text()
		arr := strings.Fields(title)
		dayName := arr[0]
		date := arr[1]

		var meals []model.Meal
		s.Find(".item").Each(func(j int, sj *goquery.Selection) {

			description := sj.Find(".title").Text()

			if description == "" {
				description = sj.Find(".text").Text()
			}

			description = strings.TrimSpace(description)
			space := regexp.MustCompile(`\s+`)
			description = space.ReplaceAllString(description, " ")

			if len(description) > 50 {
				description = description[:49]
			}

			price := strings.TrimSpace(sj.Find(".price").Text())

			meals = append(meals, model.Meal{
				Description: description + " " + price + " Kc",
			})
		})

		dayOffer = append(dayOffer,
			model.Day{
				DayName: dayName,
				Date:    date,
				Meals:   meals,
			})

	})

	restaurant := &model.Restaurant{
		Title: a.title,
		Days:  dayOffer,
	}

	return restaurant, nil
}

func New() *suzies {
	return &suzies{
		url:   "http://www.suzies.cz/poledni-menu.html",
		title: "SUZIES STEAK PUB",
	}
}

File: nodejs/restaurants/Suzies.js
const Fetch = require('./abstract/Fetch');
const Restaurant = require('../model/Restaurant');
const HTMLParser = require('node-html-parser');

class Suzies extends Fetch {
  constructor(renderFunc) {
    super('http://www.suzies.cz/poledni-menu.html');
    this.name = 'SUZIES STEAK PUB';
    this.renderFunc = renderFunc;
  }

  parseToRestaurant() {
    if (this.body === null) {
      console.log('cannot parse null');
    }

    const restaurant = new Restaurant(this.name);
    const root = HTMLParser.parse(this.body);
    const rows = root.querySelectorAll('#weekly-menu .day');

    rows.map((item) => {
      const row = HTMLParser.parse(item.childNodes);
      const title = row.querySelector('h4').text;
      const [dayName, date] = title.split(' ');

      const food = row.querySelectorAll('.item');

      const dayOffer = [];

      food.map((item2) => {
        const row2 = HTMLParser.parse(item2.childNodes);

        let descriptionElement = row2.querySelector('.title');

        if (!descriptionElement) {
          descriptionElement = row2.querySelector('.text');
        }

        let description =
          (descriptionElement && descriptionElement.text) || '';

        description = description.replace(/\s/g, ' ').trim()

        if (description.length > 50) {
          description = description.substring(0, 49) + "...";
        }

        const priceElement = row2.querySelector('.price');
        const price = (priceElement && priceElement.text) || '';
        dayOffer.push({
          description:
            description.replace(/\s/g, ' ').trim() + '  ' + price.trim() + " Kc",
        });
      });

      restaurant.pushDay(dayName, date, dayOffer);
    });

    return restaurant;
  }

  async process() {
    await this.fetch();
    const restaurant = this.parseToRestaurant();
    this.renderFunc(restaurant);
  }
}

module.exports = Suzies;
